src/storage/raw/raw_mvcc.rs

```diff
@@ -0,0 +1,244 @@
+    }
+
+    #[inline]
+    fn upper_bound(&self) -> Option<&[u8]> {
+        self.snap.upper_bound()
+    }
+
+    fn ext(&self) -> S::Ext<'_> {
+        self.snap.ext()
+    }
+}
+
+pub struct RawMvccIterator<I: Iterator> {
+    inner: I,
+    cur_key: Option<Vec<u8>>,
+    cur_value: Option<Vec<u8>>,
+    is_valid: Option<bool>,
+    is_forward: bool,
+}
+
+fn is_user_key_eq(left: &[u8], right: &[u8]) -> bool {
+    let len = left.len();
+    if len != right.len() {
+        return false;
+    }
+    Key::is_user_key_eq(left, &right[..len - U64_SIZE])
+}
+
+impl<I: Iterator> RawMvccIterator<I> {
+    fn new(inner: I) -> Self {
+        RawMvccIterator {
+            inner,
+            cur_key: None,
+            cur_value: None,
+            is_valid: None,
+            is_forward: true,
+        }
+    }
+
+    fn update_cur_kv(&mut self) {
+        if let Some(ref mut key) = self.cur_key {
+            key.clear();
+            key.extend_from_slice(self.inner.key());
+            if key.capacity() > key.len() * 2 && key.capacity() > VEC_SHRINK_THRESHOLD {
+                key.shrink_to(key.len() * 2);
+            }
+        } else {
+            self.cur_key = Some(Vec::from(self.inner.key()));
+        };
+        if let Some(ref mut value) = self.cur_value {
+            value.clear();
+            value.extend_from_slice(self.inner.value());
+            if value.capacity() > value.len() * 2 && value.capacity() > VEC_SHRINK_THRESHOLD {
+                value.shrink_to(value.len() * 2);
+            }
+        } else {
+            self.cur_value = Some(Vec::from(self.inner.value()));
+        };
+        self.is_valid = Some(true);
+    }
+
+    fn clear_cur_kv(&mut self) {
+        self.cur_key = None;
+        self.cur_value = None;
+        self.is_valid = None;
+    }
+
+    fn move_to_prev_max_ts(&mut self) -> Result<bool> {
+        if self.inner.valid()? {
+            self.update_cur_kv();
+            self.inner.prev()?;
+        } else {
+            self.clear_cur_kv();
+            return Ok(false);
+        }
+        while self.inner.valid()? {
+            // cur_key should not be None here.
+            if is_user_key_eq(self.cur_key.as_ref().unwrap(), self.inner.key()) {
+                self.update_cur_kv();
+                self.inner.prev()?;
+            } else {
+                break;
+            }
+        }
+        Ok(true)
+    }
+}
+
+// RawMvccIterator always return the latest ts of user key.
+// ts is desc encoded after user key, so it's placed the first one for the same user key.
+// Only one-way direction scan is supported. Like `seek` then `next` or `seek_for_prev` then `prev`
+impl<I: Iterator> Iterator for RawMvccIterator<I> {
+    fn next(&mut self) -> Result<bool> {
+        if !self.is_forward {
+            return Err(Error::from(ErrorInner::Other(Box::from(
+                "invalid raw mvcc operation",
+            ))));
+        }
+        let cur_key = self.inner.key().to_owned();
```

Comment:
Probably need to ensure it's valid before calling key.

Reply:
i think it is caller's responsibility to check validition before calling next().

Reply:
It's hard. That's why we have features "nortcheck" in a lot of components.